contributors/design-proposals/cpu-affinity-numa-awareness.md

```diff
@@ -0,0 +1,216 @@
+**Author**: Derek Carr
+
+**Last** **Updated**: 12/12/2016
+
+**Status**: Pending Review
+
+# CPU Affinity and NUMA Topology Awareness
+
+This proposal describes enhancements to the Kubernetes API to improve the
+utilization of compute resources for containers that have a need to avoid
+cross NUMA node memory access by containers.
+
+This proposal describes the set of changes recommended to support NUMA
+awareness in the `Node` and `Pod` API to support this feature.  It does not
+yet prescribe specifically the actor that performs NUMA aware scheduling
+decisions, instead it focuses purely on the underlying primitives required
+to support building more advanced scheduling capabilities.
+
+## What is NUMA?
+
+Non-uniform memory architecture (NUMA) describes multi-socket machines that
+subdivide memory into nodes where each node is associated with a list of
+CPU cores.  This architecture is the norm for modern machines.
+
+An interconnect bus provides connections between nodes so each CPU can
+access all memory.  The interconnect can be overwhelmed by concurrent
+cross-node traffic, and as a result, processes that need to access memory
+on a different node can experience increased latency.
+
+As a result, many applications see a performance benefit when the workload
+is affined to a particular NUMA node and CPU core(s).
+
+## NUMA topology
+
+In order to support NUMA affined workloads, the `Node` must make its
+NUMA topology available for introspection by other agents that schedule
+pods.
+
+This proposal recommends that the `NodeStatus` is augmented as follows:
+
+```
+// NodeStatus is information about the current status of a node.
+type NodeStatus struct {
+    ...
+	// Topology represents the NUMA topology of a node to aid NUMA aware scheduling.
+	// +optional
+	Topology NUMATopology
+}
+
+// NUMATopology describes the NUMA topology of a node.
+type NUMATopology struct {
+	// NUMANodes represents the list of NUMA nodes in the topology.
+	NUMANodes []NUMANode
+}
+
+// NUMANode describes a single NUMA node.
+type NUMANode struct {
+	// Identifies a NUMA node on a single host.
+	NUMANodeID string
+	// Capacity represents the total resources associated to the NUMA node.
+	// cpu: 4 <number of cores>
+	// memory: <amount of memory in normal page size>
+	// hugepages: <amount of memory in huge page size>
```

Comment:
This needs much more work to correctly future proof it to cope well with huge pages. Historically x86 only had 1 huge page size (2MB or 4MB depending on kernel address mode), but these days you can also have 1 GB huge pages. On PPC architecture there are a huge number of page sizes available. Any single host can have multiple page sizes available concurrently.

So memory needs to be represented as a struct in its own right eg something more like

type PageInfo struct {
    PageSize uint
    Count uint64
    Used uint64
}

Then the node would have an array of info   Memory []PageInfo

The 'cpu: 4' count seems redundant given that you have a CPUSet string below from which you could easily determine the count if represented as an array instead of string.